containers/java/.devcontainer/base.Dockerfile

```
@@ -26,18 +28,26 @@ ARG INSTALL_GRADLE="false"
 ARG GRADLE_VERSION=""
 ENV SDKMAN_DIR="/usr/local/sdkman"
 ENV PATH="${SDKMAN_DIR}/candidates/java/current/bin:${PATH}:${SDKMAN_DIR}/candidates/maven/current/bin:${SDKMAN_DIR}/candidates/gradle/current/bin"
-RUN bash /tmp/library-scripts/java-debian.sh "none" "${SDKMAN_DIR}" "${USERNAME}" "true" \
-    && if [ "${INSTALL_MAVEN}" = "true" ]; then bash /tmp/library-scripts/maven-debian.sh "${MAVEN_VERSION:-latest}" "${SDKMAN_DIR}" ${USERNAME} "true"; fi \
-    && if [ "${INSTALL_GRADLE}" = "true" ]; then bash /tmp/library-scripts/gradle-debian.sh "${GRADLE_VERSION:-latest}" "${SDKMAN_DIR}" ${USERNAME} "true"; fi \
-    && apt-get clean -y && rm -rf /var/lib/apt/lists/*
+# Install Java tools such as JDK, Maven and Gradle
+RUN bash /tmp/library-scripts/java-debian.sh $TARGET_JAVA_VERSION "${SDKMAN_DIR}" "${USERNAME}" "true" \
+	&& if [ "${INSTALL_MAVEN}" = "true" ]; then bash /tmp/library-scripts/maven-debian.sh "${MAVEN_VERSION:-latest}" "${SDKMAN_DIR}" ${USERNAME} "true"; fi \
+	&& if [ "${INSTALL_GRADLE}" = "true" ]; then bash /tmp/library-scripts/gradle-debian.sh "${GRADLE_VERSION:-latest}" "${SDKMAN_DIR}" ${USERNAME} "true"; fi \
+	&& apt-get clean -y && rm -rf /var/lib/apt/lists/*
+
+# Add symlinks to the installed JDK
```

```suggestion
ARG TARGET_JAVA_VERSION=17
ARG BASE_IMAGE_VERSION_CODENAME=bullseye
FROM debian:${BASE_IMAGE_VERSION_CODENAME}

ARG TARGET_JAVA_VERSION
ENV LANG en_US.UTF-8
ENV JAVA_HOME /usr/lib/jvm/msopenjdk-${TARGET_JAVA_VERSION}-amd64
ENV PATH "${JAVA_HOME}/bin:${PATH}"
COPY --from=mcr.microsoft.com/openjdk/jdk:17-ubuntu ${JAVA_HOME} ${JAVA_HOME}
```

This is coming from [MS docs](https://docs.microsoft.com/en-us/java/openjdk/containers#use-a-different-base-os-image)

---

Since we need to support arm64 architecture as well, we can‘t copy amd64 JDK from the mcr JDK using hardcoding. The tool like sdkman supports installing native arm64 JDKs.

---

@testforstephen but aren't Linux containers on macOS M1 laptops running as regular x64 containers?

---

https://github.com/microsoft/vscode-dev-containers/blob/36d591b3df00e3d5277c8d806bc57ac28293b5e1/containers/java/definition-manifest.json#L25-L29

The manifest defines multiple architectures support for Java images, the build pipeline will build docker images with buildx for multiple architecture support.

Add @Chuxel for more information about the underlying build pipeline of vscode-dev-containers.

---

@brunoborges One question for you, does your team publish arm64 JDK to sdkman?

sdkman supports native arm64 JDKs since 5.10.0, see https://github.com/sdkman/sdkman-cli/issues/830#issuecomment-766328300. The sdkman integrated in Java template is the latest version 5.13.1.

---

It is a 3rd-party package manager we do not manage.

---

oh, I see, thanks for clarification. I previously thought sdkman's MS OpenJDK was maintained by you guys.

---

@brunoborges @testforstephen No containers on Apple Silicon macs default to arm64. Docker Desktop can emulate x86, but this is slow. We do build an arm64 version of the bullseye variant for this reason. They get uploaded with the same identifier and the architecture is auto selected.

Is there not an arm64 base image variant for MS JDK? With the recent shifts to MacBook Pros, there's been quite a bit of interest in it.

---

Yes, we do provide JDK for Linux ARM.

But I'm not sure we have arm packages published on packages.microsoft.com.

The tar.gz can be downloaded from the website though, and simply extracted (and configured).

---

The latest commit uses Microsoft's download url to get MS OpenJDK tar.gz file and install it on the image.

---

Nice - Is there a sha or signature we should verify for the download?  It looks like the name of the package could be used to get the sha/sig.  It doesn't seem like there's one for just `17`, but this exists:

```
https://aka.ms/download-jdk/microsoft-jdk-17.0.1.12.1-linux-x64.tar.gz.sig
```

---

Appending `.sha256sum.txt` to the URL  can download the sha256 checksum file.

https://aka.ms/download-jdk/microsoft-jdk-17-linux-x64.tar.gz.sha256sum.txt

It's coming from [MS Docs](https://docs.microsoft.com/en-us/java/openjdk/download-major-urls#checksums).